src/audio/codec_adapter/codec/dts.c

```diff
@@ -0,0 +1,365 @@
+				sdk_version.build);
+		}
+	}
+
+	if (ret)
+		comp_err(dev, "dts_codec_init() failed %d %d", ret, dts_result);
+
+	comp_dbg(dev, "dts_codec_init() done");
+
+	return ret;
+}
+
+int dts_codec_prepare(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	DtsSofInterfaceBufferConfiguration buffer_configuration;
+	DtsSofInterfaceResult dts_result;
+
+	comp_dbg(dev, "dts_codec_prepare() start");
+
+	ret = dts_effect_populate_buffer_configuration(dev, &buffer_configuration);
+	if (ret) {
+		comp_err(dev,
+			"dts_codec_prepare() dts_effect_populate_buffer_configuration failed %d",
+			ret);
+		return ret;
+	}
+
+	dts_result = dtsSofInterfacePrepare(
+		(DtsSofInterfaceInst *)codec->private,
+		&buffer_configuration,
+		&codec->cpd.in_buff,
+		&codec->cpd.in_buff_size,
+		&codec->cpd.out_buff,
+		&codec->cpd.out_buff_size);
+	ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+
+	if (ret)
+		comp_err(dev, "dts_codec_prepare() failed %d", ret);
+
+	comp_dbg(dev, "dts_codec_prepare() done");
+
+	return ret;
+}
+
+int dts_codec_init_process(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	DtsSofInterfaceResult dts_result;
+
+	comp_dbg(dev, "dts_codec_init_process() start");
+
+	dts_result = dtsSofInterfaceInitProcess(codec->private);
+	ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+
+	codec->cpd.produced = 0;
+	codec->cpd.consumed = 0;
+	codec->cpd.init_done = 1;
+
+	if (ret)
+		comp_err(dev, "dts_codec_init_process() failed %d %d", ret, dts_result);
+
+	comp_dbg(dev, "dts_codec_init_process() done");
+
+	return ret;
+}
+
+int dts_codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	DtsSofInterfaceResult dts_result;
+	unsigned int bytes_processed = 0;
+
+	comp_dbg(dev, "dts_codec_process() start");
+
+	dts_result = dtsSofInterfaceProcess(codec->private, &bytes_processed);
+	ret = dts_effect_convert_sof_interface_result(dev, dts_result);
+
+	codec->cpd.consumed = !ret ? bytes_processed : 0;
+	codec->cpd.produced = !ret ? bytes_processed : 0;
+
+	if (ret)
+		comp_err(dev, "dts_codec_process() failed %d %d", ret, dts_result);
+
+	comp_dbg(dev, "dts_codec_process() done");
+
+	return ret;
+}
+
+int dts_codec_apply_config(struct comp_dev *dev)
+{
+	int ret = 0;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct codec_config *config;
+	struct codec_param *param;
+	uint32_t config_header_size;
+	uint32_t config_data_size;
+	uint32_t param_header_size;
+	uint32_t param_data_size;
+	uint32_t i;
+	uint32_t param_number = 0;
+	DtsSofInterfaceResult dts_result;
+
+	comp_dbg(dev, "dts_codec_apply_config() start");
+
+	config = &codec->r_cfg;
+
+	/* Check that config->data isn't invalid and has size greater than 0 */
+	config_header_size = sizeof(config->size) + sizeof(config->avail);
+	if (config->size < config_header_size) {
+		comp_dbg(dev, "dts_codec_apply_config() config->data is invalid");
```

Comment:
this sounds like it should be an error

Reply:
Whilst we possibly could recover, I think it's safest here to return `-EINVAL` at the first sign of an individual config blob failing to be applied in our codec. Changed code to early out with error.